### PR TITLE
Ddcnls 5118:Ated UI tests

### PR DESCRIPTION
**Pull request by ashwinirudrappa**

### CODE REVIEW

**Comment by Gaur4vGaur** [comment]
can you please provide description to PR

File: test/forms/PropertyDetailsTaxAvoidanceSpec.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+package forms
+
+import forms.PropertyDetailsForms.{propertyDetailsTaxAvoidanceForm, validatePropertyDetailsTaxAvoidance}
+import org.scalatest.Matchers
+import org.scalatestplus.play.OneAppPerSuite
+import play.api.i18n.{Messages, MessagesApi}
+import play.api.test.FakeRequest
+import uk.gov.hmrc.play.test.UnitSpec
+
+class PropertyDetailsTaxAvoidanceSpec  extends UnitSpec with Matchers with OneAppPerSuite {
```
Comment (Gaur4vGaur) [inline test/forms/PropertyDetailsTaxAvoidanceSpec.scala]: please use playspec

File: test/forms/PropertyDetailsRevaluedSpec.scala [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+          "revaluedValue" -> "150000000",
+          "revaluedDate.day" -> "12",
+          "revaluedDate.month" -> "",
+          "revaluedDate.year" -> ""
+        )
+
+        PropertyDetailsForms.validatePropertyDetailsRevalued(periodKey,  propertyDetailsRevaluedForm.bind(input)).fold(
+          hasErrors => {
+            hasErrors.errors.length shouldBe  2
+            println(hasErrors.errors)
```
Comment (Gaur4vGaur) [inline test/forms/PropertyDetailsRevaluedSpec.scala]: please get rid of println

File: test/forms/PropertyFormValidationSpec.scala [hunk 3]
```diff
@@ -17,17 +17,22 @@
...
 import forms.PropertyDetailsForms._
 import org.joda.time.LocalDate
 import uk.gov.hmrc.play.test.UnitSpec
 import org.scalatest.Matchers
 import org.scalatestplus.play.OneServerPerSuite
 import play.api.data.FormError
+import play.api.i18n.{Messages, MessagesApi}
+import play.api.test.FakeRequest
 
 class PropertyFormValidationSpec extends UnitSpec with Matchers with OneServerPerSuite {
```
Comment (Gaur4vGaur) [inline test/forms/PropertyFormValidationSpec.scala]: please use play spec

File: test/forms/PropertyFormValidationSpec.scala [hunk 4]
```diff
@@ -17,17 +17,22 @@
...
 
-  "PropertDetailsForm" should {
+  implicit lazy val messagesApi = app.injector.instanceOf[MessagesApi]
+  implicit lazy val messages = messagesApi.preferred(FakeRequest())
+
+  "PropertyDetailsForm" should {
```
Comment (Gaur4vGaur) [inline test/forms/PropertyFormValidationSpec.scala]: please use must for consistency

File: test/utils/viewHelpers/AtedViewSpec.scala [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+import org.scalatestplus.play.{OneAppPerSuite, PlaySpec}
+import play.api.{Application, Mode}
+import play.api.i18n.Messages
+import play.api.inject.guice.{GuiceApplicationBuilder, GuiceableModule}
+import play.api.test.FakeRequest
+import play.twirl.api.Html
+
+trait AtedViewSpec extends PlaySpec
+  with JsoupMatchers
+  with OneAppPerSuite {
```
Comment (Gaur4vGaur) [inline test/utils/viewHelpers/AtedViewSpec.scala]: why have changed from OneServerPerSuite?

File: test/utils/viewHelpers/AtedViewSpec.scala [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+  protected val bindModules: Seq[GuiceableModule] = Seq()
+
+  implicit override lazy val app: Application = new GuiceApplicationBuilder()
+    .disable[com.kenshoo.play.metrics.PlayModule]
+    .bindings(bindModules:_*).in(Mode.Test)
+    .build()
```
Comment (Gaur4vGaur) [inline test/utils/viewHelpers/AtedViewSpec.scala]: I dont think we need to disable metrics module.

File: test/views/propertyDetails/html/isFullTaxPeriodSpec.scala [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+package views.propertyDetails.html
+
+import forms.PropertyDetailsForms
+import org.joda.time.LocalDate
+import play.api.data.FormError
+import play.twirl.api.Html
+import utils.viewHelpers.AtedViewSpec
+
+class isFullTaxPeriodSpec extends AtedViewSpec {
```
Comment (Gaur4vGaur) [inline test/views/propertyDetails/html/isFullTaxPeriodSpec.scala]: can you move the test to html package please

File: test/views/propertyDetails/html/propertyDetailsAcquisitionSpec.scala [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+package views.propertyDetails.html
+
+import forms.PropertyDetailsForms
+import play.api.data.{Form, FormError}
+import play.twirl.api.Html
+import utils.viewHelpers.AtedViewSpec
+
+class propertyDetailsAcquisitionSpec extends AtedViewSpec {
```
Comment (Gaur4vGaur) [inline test/views/propertyDetails/html/propertyDetailsAcquisitionSpec.scala]: can you move the test to html package please

File: test/views/propertyDetails/html/propertyDetailsProfessionallyValuedSpec.scala [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package views.propertyDetails.html
+
+import forms.PropertyDetailsForms
+import play.api.data.{Form, FormError}
+import play.twirl.api.Html
+import utils.viewHelpers.AtedViewSpec
+
+class propertyDetailsProfessionallyValuedSpec extends AtedViewSpec {
```
Comment (Gaur4vGaur) [inline test/views/propertyDetails/html/propertyDetailsProfessionallyValuedSpec.scala]: can you move the test to html package please

File: test/views/propertyDetails/html/propertyDetailsRevaluedSpec.scala [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package views.propertyDetails.html
+
+import forms.PropertyDetailsForms
+import play.api.data.{Form, FormError}
+import play.twirl.api.Html
+import utils.viewHelpers.AtedViewSpec
+
+class propertyDetailsRevaluedSpec extends AtedViewSpec {
```
Comment (Gaur4vGaur) [inline test/views/propertyDetails/html/propertyDetailsRevaluedSpec.scala]: can you move the test to html package please

File: test/views/propertyDetails/html/propertyDetailsTaxAvoidanceSpec.scala [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+ */
+
+package views.propertyDetails.html
+
+import forms.PropertyDetailsForms
+import play.api.data.{Form, FormError}
+import play.twirl.api.Html
+import utils.viewHelpers.AtedViewSpec
+
+class propertyDetailsTaxAvoidanceSpec extends AtedViewSpec {
```
Comment (Gaur4vGaur) [inline test/views/propertyDetails/html/propertyDetailsTaxAvoidanceSpec.scala]: can you move the test to html package please

File: app/forms/PropertyDetailsFormsValidation.scala [hunk 12]
```diff
@@ -35,16 +35,16 @@ object PropertyDetailsFormsValidation {
     def validateAvoidanceScheme(avoidanceSchemeNo: Option[String]): Seq[Option[FormError]] = {
       avoidanceSchemeNo.getOrElse("") match {
         case a if a.isEmpty => Seq(Some(FormError("taxAvoidanceScheme", Messages("ated.property-details-period.taxAvoidanceScheme.error.empty"))))
-        case a if a.length != 8 => Seq(Some(FormError("taxAvoidanceScheme", Messages("ated.property-details-period.taxAvoidanceScheme.error.wrong-length"))))
         case a if Try(a.toInt).isFailure => Seq(Some(FormError("taxAvoidanceScheme", Messages("ated.property-details-period.taxAvoidanceScheme.error.numbers"))))
+        case a if a.length != 8 => Seq(Some(FormError("taxAvoidanceScheme", Messages("ated.property-details-period.taxAvoidanceScheme.error.wrong-length"))))
```
[the remark this answers is not in the document]
Comment (ashwinirudrappa) [reply app/forms/PropertyDetailsFormsValidation.scala]: Yes,Its a good reason it was never entering this block when we enter text as string and it was always throwing error message as invalid length where as message suppose to be enter only number

File: app/forms/PropertyDetailsFormsValidation.scala [hunk 13]
```diff
@@ -35,16 +35,16 @@ object PropertyDetailsFormsValidation {
...
         case _ => Seq(None)
       }
     }
     def validatePromoterReference(promoterReference: Option[String]): Seq[Option[FormError]] = {
       promoterReference.getOrElse("") match {
         case a if a.isEmpty => Seq(Some(FormError("taxAvoidancePromoterReference", Messages("ated.property-details-period.taxAvoidancePromoterReference.error.empty"))))
-        case a if a.length != 8 => Seq(Some(FormError("taxAvoidancePromoterReference", Messages("ated.property-details-period.taxAvoidancePromoterReference.error.wrong-length"))))
         case a if Try(a.toInt).isFailure => Seq(Some(FormError("taxAvoidancePromoterReference", Messages("ated.property-details-period.taxAvoidancePromoterReference.error.numbers"))))
+        case a if a.length != 8 => Seq(Some(FormError("taxAvoidancePromoterReference", Messages("ated.property-details-period.taxAvoidancePromoterReference.error.wrong-length"))))
```
[the remark this answers is not in the document]
Comment (ashwinirudrappa) [reply app/forms/PropertyDetailsFormsValidation.scala]: Yes,Its a good reason it was never entering this block when we enter text as string and it was always throwing error message as invalid length where as message suppose to be enter only number

File: test/controllers/editLiability/BankDetailsControllerSpec.scala [hunk 14]
```diff
@@ -112,6 +112,7 @@ class BankDetailsControllerSpec extends PlaySpec with OneServerPerSuite with Moc
         saveWithAuthorisedUser(inputJson) {
           result =>
             status(result) must be(BAD_REQUEST)
+            contentAsString(result) must include("You must answer the UK bank account question")
```
[the remark this answers is not in the document]
Comment (ashwinirudrappa) [reply test/controllers/editLiability/BankDetailsControllerSpec.scala]: It is in unit test but added here as wel it was checking only BadRequest

File: test/views/propertyDetails/html/isFullTaxPeriodSpec.scala (hunk 7 above)
Comment (ashwinirudrappa) [reply test/views/propertyDetails/html/isFullTaxPeriodSpec.scala]: it is in html package

File: test/forms/PropertyDetailsRevaluedSpec.scala [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+import forms.PropertyDetailsForms.{propertyDetailsRevaluedForm, propertyDetailsTaxAvoidanceForm, validatePropertyDetailsTaxAvoidance}
+import org.joda.time.LocalDate
+import org.scalatest.Matchers
+import org.scalatestplus.play.OneAppPerSuite
+import play.api.i18n.{Messages, MessagesApi}
+import play.api.test.FakeRequest
+import uk.gov.hmrc.play.test.UnitSpec
+
+class PropertyDetailsRevaluedSpec extends UnitSpec with Matchers with OneAppPerSuite {
```
[the remark this answers is not in the document]
Comment (ashwinirudrappa) [reply test/forms/PropertyDetailsRevaluedSpec.scala]: Fixed it

File: test/controllers/propertyDetails/IsFullTaxPeriodControllerSpec.scala [hunk 16]
```diff
@@ -184,6 +174,7 @@ class IsFullTaxPeriodControllerSpec extends PlaySpec with OneServerPerSuite with
           submitWithAuthorisedUser(inputJson) {
             result =>
               status(result) must be(BAD_REQUEST)
+              contentAsString(result) must include("There is a problem with this question")
```
[the remark this answers is not in the document]
Comment (ashwinirudrappa) [reply test/controllers/propertyDetails/IsFullTaxPeriodControllerSpec.scala]: Fixed it

File: test/views/propertyDetails/html/propertyDetailsAcquisitionSpec.scala (hunk 8 above)
Comment (ashwinirudrappa) [reply test/views/propertyDetails/html/propertyDetailsAcquisitionSpec.scala]: Moved it

File: test/views/propertyDetails/html/propertyDetailsProfessionallyValuedSpec.scala (hunk 9 above)
Comment (ashwinirudrappa) [reply test/views/propertyDetails/html/propertyDetailsProfessionallyValuedSpec.scala]: Moved it

File: test/views/propertyDetails/html/propertyDetailsRevaluedSpec.scala (hunk 10 above)
Comment (ashwinirudrappa) [reply test/views/propertyDetails/html/propertyDetailsRevaluedSpec.scala]: Moved it

File: test/views/propertyDetails/html/propertyDetailsTaxAvoidanceSpec.scala (hunk 11 above)
Comment (ashwinirudrappa) [reply test/views/propertyDetails/html/propertyDetailsTaxAvoidanceSpec.scala]: Moved it

File: test/utils/viewHelpers/AtedViewSpec.scala (hunk 5 above)
Comment (ashwinirudrappa) [reply test/utils/viewHelpers/AtedViewSpec.scala]: OneServerPerSuite is deprecated so using GuiceOneAppPerSuite

File: app/forms/PropertyDetailsFormsValidation.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (Gaur4vGaur) [reply app/forms/PropertyDetailsFormsValidation.scala]: The ticket is to improve the code coverage. If we think it is a bug can we please discuss it with Bernadette.

File: app/forms/PropertyDetailsFormsValidation.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (Gaur4vGaur) [reply app/forms/PropertyDetailsFormsValidation.scala]: The ticket is to improve the code coverage. If we think it is a bug can we please discuss it with Bernadette.

File: test/forms/PropertyFormValidationSpec.scala [hunk 17]
```diff
@@ -17,33 +17,38 @@
...
-import org.scalatestplus.play.OneServerPerSuite
+import org.scalatest.MustMatchers
+import org.scalatestplus.play.PlaySpec
+import org.scalatestplus.play.guice.GuiceOneServerPerSuite
 import play.api.data.FormError
+import play.api.i18n.{Messages, MessagesApi}
+import play.api.test.FakeRequest
 
-class PropertyFormValidationSpec extends UnitSpec with Matchers with OneServerPerSuite {
+class PropertyFormValidationSpec extends PlaySpec with MustMatchers with GuiceOneServerPerSuite {
```
Comment (Gaur4vGaur) [inline test/forms/PropertyFormValidationSpec.scala]: 👍

File: test/views/html/editLiability/EditLiabilityDatesLiableSpec.scala [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+      errorDoc.getElementsMatchingOwnText(messages("ated.property-details-value.startDate.error.empty")).hasText mustBe true
+      errorDoc.getElementsMatchingOwnText(messages("ated.property-details-period.datesLiable.general.error.startDate")).hasText mustBe true
+      errorDoc.getElementsMatchingOwnText(messages("ated.property-details-value.endDate.error.empty")).hasText mustBe true
+      errorDoc.getElementsMatchingOwnText(messages("ated.property-details-period.datesLiable.general.error.endDate")).hasText mustBe true
+
+      errorDoc.getElementsMatchingOwnText(messages("ated.property-details-period.change-dates-liable.startDate")).hasText mustBe true
+      errorDoc.getElementsMatchingOwnText(messages("ated.property-details-period.datesLiable.startDate.hint")).hasText mustBe true
+      errorDoc.getElementsMatchingOwnText(messages("ated.property-details-period.change-dates-liable.endDate")).hasText mustBe true
+      errorDoc.getElementsMatchingOwnText(messages("ated.property-details-period.datesLiable.endDate.hint")).hasText mustBe true
```
Comment (Gaur4vGaur) [inline test/views/html/editLiability/EditLiabilityDatesLiableSpec.scala]: can you please use haveErrorSummary and haveErrorNotification from JsoupMatchers. It checks for both error Text and position of errors

File: test/views/html/propertyDetails/PropertyDetailsTaxAvoidanceSpec.scala [hunk 19]
```diff
@@ -68,9 +68,14 @@ class PropertyDetailsTaxAvoidanceSpec extends AtedViewSpec {
       errorDoc.getElementsMatchingOwnText(messages("ated.property-details-period.taxAvoidancePromoterReference.error.empty")).hasText mustBe true
       errorDoc.getElementsMatchingOwnText(messages("ated.property-details-period-error.general.taxAvoidancePromoterReference")).hasText mustBe true
     }
+
+    "check error message for empty form" in {
+      doc.getElementsContainingOwnText(messages("ated.property-details-period-error.general.isFullPeriod")).hasText mustBe true
+      doc.getElementsContainingOwnText(messages("ated.property-details-period.isTaxAvoidance.error-field-name")).hasText mustBe true
```
Comment (Gaur4vGaur) [inline test/views/html/propertyDetails/PropertyDetailsTaxAvoidanceSpec.scala]: can you please use haveErrorSummary and haveErrorNotification from JsoupMatchers. It checks for both error Text and position of errors

File: test/views/html/editLiability/EditLiabilityDatesLiableSpec.scala (hunk 18 above)
Comment (ashwinirudrappa) [reply test/views/html/editLiability/EditLiabilityDatesLiableSpec.scala]: Fixed it

File: test/views/html/propertyDetails/PropertyDetailsTaxAvoidanceSpec.scala (hunk 19 above)
Comment (ashwinirudrappa) [reply test/views/html/propertyDetails/PropertyDetailsTaxAvoidanceSpec.scala]: Fixed it

**Pull request merged**